File: libs/imagine/src/algs/ops/integers.cs
//-----------------------------------------------------------------------------
// Copyright   :  (c) Chris Moore, 2020
// License     :  MIT
//-----------------------------------------------------------------------------
namespace Z0
{
    using static System.Runtime.CompilerServices.Unsafe;

    partial class Algs
    {
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref sbyte int8<T>(ref T src)
            => ref As<T,sbyte>(ref src);

        [MethodImpl(Inline), Op, Closures(AllNumeric)]
        public static ref short int16<T>(ref T src)
            => ref As<T,short>(ref src);

        [MethodImpl(Inline), Op, Closures(AllNumeric)]
        public static ref int int32<T>(ref T src)
            => ref As<T,int>(ref src);

        [MethodImpl(Inline), Op, Closures(AllNumeric)]
        public static ref long int64<T>(ref T src)
            => ref As<T,long>(ref src);

        /// <summary>
        /// Presents a T-references as a <see cref='byte'/> reference
        /// </summary>
        /// <param name="src">The source reference</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref byte u8<T>(in T src)
            => ref @as<T,byte>(src);

        /// <summary>
        /// Presents a T-reference as a <see cref='ushort'/> reference
        /// </summary>
        /// <param name="src">The source reference</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref ushort u16<T>(in T src)
            => ref @as<T,ushort>(src);

        /// <summary>
        /// Presents a parametric references as a <see cref='uint'/> reference
        /// </summary>
        /// <param name="src">The source reference</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref uint u32<T>(in T src)
            => ref @as<T,uint>(src);

        /// <summary>
        /// Presents a T-references as a <see cref='ulong'/> reference
        /// </summary>
        /// <param name="src">The source reference</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref ulong u64<T>(in T src)
            => ref @as<T,ulong>(src);

        /// <summary>
        /// Presents a <typeparamref name='T'/> reference as an <see cref='sbyte'/> reference
        /// </summary>
        /// <param name="src">The source reference</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref sbyte i8<T>(in T src)
            => ref @as<T,sbyte>(src);

        /// <summary>
        /// Presents a T-references as a <see cref='short'/> reference
        /// </summary>
        /// <param name="src">The source reference</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref short i16<T>(in T src)
            => ref @as<T,short>(src);

        /// <summary>
        /// Presents a T-references as a <see cref='int'/> reference
        /// </summary>
        /// <param name="src">The source reference</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref int i32<T>(in T src)
            => ref @as<T,int>(src);

        /// <summary>
        /// Presents a T-references as a <see cref='long'/> reference
        /// </summary>
        /// <param name="src">The source reference</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref long i64<T>(in T src)
            => ref @as<T,long>(src);

        /// <summary>
        /// Presents a parametric source reference to a <see cref='byte'/>
        /// </summary>
        /// <param name="src">The source value</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref byte uint8<T>(ref T src)
            => ref As<T,byte>(ref src);

        /// <summary>
        /// Presents a parametric source reference to a <see cref='ushort'/>
        /// </summary>
        /// <param name="src">The source value</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref ushort uint16<T>(ref T src)
            => ref As<T,ushort>(ref src);

        /// <summary>
        /// Presents a parametric source reference to a <see cref='uint'/>
        /// </summary>
        /// <param name="src">The source value</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref uint uint32<T>(ref T src)
            => ref As<T,uint>(ref src);

        /// <summary>
        /// Presents a parametric source reference to a <see cref='ulong'/>
        /// </summary>
        /// <param name="src">The source value</param>
        /// <typeparam name="T">The source type</typeparam>
        [MethodImpl(Inline), Op, Closures(Closure)]
        public static ref ulong uint64<T>(ref T src)
            => ref As<T,ulong>(ref src);
    }
}

File: src/tools/site/core/integers.ts
import {Nat} from "./nats"

export interface Int<N extends Nat,T> {
    width:N
    value:T
}

export type UInt<N extends Nat,T> = {
    width:N
    value:T
}

export type Sign = 0 | 1

export type SInt<N extends Nat,T> = {
    width:N
    value:T
    sign?:Sign
}

export type Sum<N extends Nat,T> = {    
    a:Int<N,T>
    b:Int<N,T>
}

export type int8<V> = SInt<8,V>
export type uint8<V> = UInt<8,V>

export type int16<T> = SInt<16,T>
export type uint16<T> = UInt<16,T>

export type int32<T> = SInt<32,T>
export type uint32<T> = UInt<32,T>

export type int64<T> = SInt<64,T>
export type uint64<T> = UInt<64,T>

export function u8<T>(value:T) : uint8<T> {
    return {
        width:8,
        value
    }
}

export function u16<T>(value:T) : uint16<T> {
    return {
        width:16,
        value
    }
}

export function u32<T>(value:T) : uint32<T> {
    return {
        width:32,
        value
    }
}

export function u64<T>(value:T) : int64<T> {
    return {
        width:64,
        value
    }
}

export function i8<T>(value:T, negative?:boolean) : int8<T> {
    return {
        width:8,
        value,
        sign:negative ? 1 : 0
    }
}

export function i16<T>(value:T, negative?:boolean) : int16<T> {
    return {
        width:16,
        value,
        sign:negative ? 1 : 0
    }
}

export function i32<T>(value:T, negative?:boolean) : int32<T> {
    return {
        width:32,
        value,
        sign:negative ? 1 : 0
    }
}

export function i64<T>(value:T, negative?:boolean) : int64<T> {
    return {
        width:64,
        value,
        sign:negative ? 1 : 0
    }
}
